config: Handle rich (non-string) types in TOML file

dbaty opened the pull request:
## Description

Before that, we had to use strings in a TOML configuration file, like
this:

```toml
enable = "use-symbolic-message-instead,useless-suppression"
jobs = "10"
suggestion-mode = "no"
```

TOML supports rich types like list, integer and boolean. They make for
a more readable and less error-prone file. We can now express the same
configuration like this:

```toml
enable = [
    "use-symbolic-message-instead",
    "useless-suppression",
]
jobs = 10
suggestion-mode = false
```

## Type of Changes
|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |


## Related Issue

Closes #3538 

## Tests

Previous tests were not quite right because they used `MagicMock`, which happily says yes to, well, whatever you throw at it, including `magic_mocked_thing.called_with("anything", "really")`. ;)


## Alternative

I aimed for the least invasive change, but I think that we may be trying too hard to keep using ConfigParser. Here we read TOML files (with rich types such as list), then convert back to strings to shoehorn the values into ConfigParser, then we deserialize again to rich types for final use. I feel like we may want to get rid of ConfigParser. Is this a goal that we would want to pursue?

Cons:
- a (limited?) risk of getting it wrong, especially for INI files which are working (aka: breaking something that has been working well for a long time, I guess).
- ?

Pros:
- a saner way to read files;
- we'll be ready when ``.pysetup.yaml`` becomes the next Standard Way (tm) to hold configuration in a Python project :trollface: 

I did not look at all at how it could look or if it's even technically feasible (if one was willing to do it).

PCManticore commented on tests/test_config.py:
```diff
@@ -1,37 +1,98 @@
+        msgid for msgid, _enabled in runner.linter._msgs_state.items() if _enabled
+    }
+    # logging-format-interpolation, logging-fstring-interpolation
+    assert enabled == {"W1202", "W1203"}
```
Rather than inspecting `_msgs_state` which should be an internal concern of the linter, you could use `runner.linter.is_message_enabled()` instead.

PCManticore commented on tests/test_config.py:
```diff
@@ -1,37 +1,98 @@
+    # logging-format-interpolation, logging-fstring-interpolation
+    assert enabled == {"W1202", "W1203"}
+    assert runner.linter.config.jobs == 10
+    type_checker = runner.linter._checkers["typecheck"][0]
```
I would prefer if we'd not check this as it depends on a private variable set in a checker. The number of jobs though should be fine.

dbaty commented:
Thanks for the quick review, @PCManticore . :)

tl;dr: I applied your recommendations, tweaked a few minor things, rebased off of master and squashed my commits. All builds are green.

Details:

1. I applied your recommendations: the tests now use `is_message_enabled()` and check for another boolean option (`reports`) that is accessible through a public API.

2. The tests now use only "disable" instead of both "disable" and "enable", which I could not make work. (First I tried keeping "enable" and setting `disable = "all"`, but then the tests would randomly fails, at least on Python 3.5, because sometimes "disable" was processed before "enable", and sometimes it was the other way around, so `_msgs_state` was populated differently.)

3. I moved the changelog entry to the new 2.5.3 section. TOML support is new in 2.5 and I think that such a bug fix belongs to a patch release. I can move it back to the 2.6 section if you disagree.
